apps/sms/js/thread_list_ui.js

```diff
@@ -449,88 +455,80 @@ var ThreadListUI = {
-          messageIdsToDelete = [],
-          threadCountToDelete = 0;
-
-      function exitEditMode() {
-        ThreadListUI.cancelEdit();
-        WaitingScreen.hide();
+    return Utils.confirm(
+      {
+        id: 'deleteThreads-confirmation-message',
+        args: { n: selected.length }
+      },
+      null,
+      {
+        text: 'delete',
+        className: 'danger'
       }
+    ).then(this.deleteThreadDraft.bind(this, selected));
+  },
 
-      function onAllThreadMessagesRetrieved() {
-        if (!--threadCountToDelete) {
-          MessageManager.deleteMessages(messageIdsToDelete);
-
-          threadIdsToDelete.forEach(function(threadId) {
-            ThreadListUI.deleteThread(threadId);
-          });
+  deleteThreadDraft: function thlui_deleteThreadDraft(selected) {
+    var threadIds = [], draftIds = [];
 
-          messageIdsToDelete = threadIdsToDelete = null;
+    WaitingScreen.show();
 
-          exitEditMode();
-        }
-      }
-
-      function onThreadMessageRetrieved(message) {
-        messageIdsToDelete.push(message.id);
-        return true;
+    selected.forEach((id) => {
+      // Coerce the threadId back to a number MobileMessageFilter and all
+      // other platform APIs expect this id to be a number.
+      var threadId = +id;
+      var isDraft = typeof Threads.get(threadId) === 'undefined';
+
+      if (isDraft) {
+        draftIds.push(threadId);
+        this.removeThread(threadId);
+      } else {
+        threadIds.push(threadId);
+        this.deleteThread(threadId);
       }
+    });
 
-      WaitingScreen.show();
-
-      threadIdsToDelete = selected.reduce(function(list, value) {
-        // Coerce the threadId back to a number MobileMessageFilter and all
-        // other platform APIs expect this value to be a number.
-        var threadId = +value;
-        var isDraft = typeof Threads.get(threadId) === 'undefined';
-
-        if (isDraft) {
-          Drafts.delete(Drafts.get(threadId));
-          ThreadListUI.removeThread(threadId);
-        } else {
-          list.push(threadId);
-        }
+    this.performDeleteThreadDraft(selected, draftIds, threadIds);
+  },
 
-        return list;
-      }, []);
+  performDeleteThreadDraft:
+    function thlui_performDeleteThreadDraft(selected, draftIds, threadIds) {
+    var threadCountToDelete = 0, messageIdsToDelete = [];
 
-      // That means that we've just removed some drafts
-      if (threadIdsToDelete.length !== selected.length) {
-        Drafts.store();
-      }
+    if(draftIds) {
+      draftIds.forEach(function(threadId) {
+        Drafts.delete(Drafts.get(threadId));
+      });
+    }
+    if (threadIds.length !== selected.length) {
+      Drafts.store();
+    }
 
-      if (!threadIdsToDelete.length) {
-        exitEditMode();
-        return;
+    var onAllThreadMessagesRetrieved = () => {
+      if (!--threadCountToDelete) {
+        MessageManager.deleteMessages(messageIdsToDelete);
+        this.cancelEdit();
```

Comment:
I think we never call cancelEdit if there is no thread to delete (means only drafts)